src/Symfony/Component/Security/Http/EventListener/CsrfProtectionListener.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Security\Core\Exception\InvalidCsrfTokenException;
+use Symfony\Component\Security\Csrf\CsrfToken;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\CsrfTokenBadge;
+use Symfony\Component\Security\Http\Event\VerifyAuthenticatorCredentialsEvent;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @final
+ * @experimental in 5.1
+ */
+class CsrfProtectionListener implements EventSubscriberInterface
+{
+    private $csrfTokenManager;
+
+    public function __construct(CsrfTokenManagerInterface $csrfTokenManager)
+    {
+        $this->csrfTokenManager = $csrfTokenManager;
+    }
+
+    public function verifyCredentials(VerifyAuthenticatorCredentialsEvent $event): void
+    {
+        $passport = $event->getPassport();
+        if (!$passport->hasBadge(CsrfTokenBadge::class)) {
+            return;
+        }
+
+        /** @var CsrfTokenBadge $badge */
+        $badge = $passport->getBadge(CsrfTokenBadge::class);
```

Comment:
In general, what do you think about adding this type of check in these listeners?

```
if ($badge->isResolved()) {
    return;
}
```

The idea would be that a user could implement a listener with a higher priority that resolves the CSRF token with custom logic (maybe a different CSRF token manager - could be an edge case, but this may apply better to other listeners).